simple-auth/src/main/java/com/fga/example/service/DocumentService.java

```diff
@@ -21,23 +23,38 @@ public DocumentService(OpenFgaClient fgaClient) {
     }
 
     /**
-     * Simple Fga bean available for use in preauthorize. TODOs:
-     * - without any auth, we are just hard-coding the userId for simple demo purposes. Can we infer it from the principal in a real example?
+     * Simple Fga bean available for use in preauthorize.
      * - multiple params of same type with optional params (e.g., userId) could be unwieldy
      */
-    @PreAuthorize("@openFga.check(#id, 'document', 'reader', 'user', '123')")
-    public String getDocumentWithSimpleFgaBean(String id) {
+    @PreAuthorize("@openFga.check(#id, 'document', 'reader', 'user', authentication?.name)")
+    public String getDocumentWithPreAuthorize(String id) {
             return "You have reader access to this document";
     }
 
+    /**
+     * Uses the new PreAuthorize meta-annotation support
+     */
+    @PreOpenFgaCheck(userType="'user'", relation="'reader'", objectType="'document'", object="#id")
+    public String getDocumentWithPreOpenFgaCheck(String id) {
+        return "You have reader access to this document";
+    }
+
+    /**
+     * Uses the new PreAuthorize meta-annotation support
+     */
+    @PreReadDocumentCheck("#id")
```

Comment:
> I'm curious about the way we might envision something like this to be used. If a pre/post Fga check meta-annotation is provided by an FGA starter or in Spring, then a customer could further extend it for specific use cases, such as shown here, is that correct?

I believe users would provide a `@PreReadDocumentCheck` annotation themselves because we don't know what relations they have in their application. 

I'm not sure I know what to do with `@PreOpenFgaCheck`, but I don't think that it belongs in Spring Security. My concerns are:

- The annotation is tied to a bean by a specific name
- A method on the bean with a specific method signature

This support needs a specific method signature. If we can be certain of a generic method signature I think that we are better off creating a generic interface (perhaps `AuthorizationManager<ReBACContext>`), a OpenFGA implementation of the interface, and then an annotation `@ReBAC` that supports SpEL and meta-annotations.

> Also curious if such a similar extension could be done today (e.g., Spring Security 6.2), while a custom AOP handler would still be needed, could a similar extension still be done?

Yes this is possible, but it is not something you could build off of the existing PreAuthorize. It would need to be built from the ground up.

Reply:
> I believe users would provide a @PreReadDocumentCheck annotation themselves because we don't know what relations they have in their application.

Ok, makes sense, good to confirm my understanding of its usage. Super cool!

> I'm not sure I know what to do with @PreOpenFgaCheck, but I don't think that it belongs in Spring Security. My concerns are:
> 
> - The annotation is tied to a bean by a specific name
> - A method on the bean with a specific method signature
>
> This support needs a specific method signature. If we can be certain of a generic method signature I think that we are better off creating a generic interface (perhaps AuthorizationManager<ReBACContext>), a OpenFGA implementation of the interface, and then an annotation @ReBAC that supports SpEL and meta-annotations.

I tend to agree with your assessment that it does not belong in Spring Security. A higher-level abstraction does make sense, one that OpenFGA could implement. There are a few other use cases I need to explore further, as the current example to perform a check does not include the ability to pass contextual tuples or context params (creating issues to explore shortly).

One of the other options we've been considering is to create an FGA Spring Boot Starter, which as we work through that could help flesh out a potential higher level abstraction.